raphtory/src/python/utils/mod.rs

```diff
@@ -17,43 +20,77 @@ use numpy::IntoPyArray;
 use pyo3::{
     exceptions::{PyRuntimeError, PyTypeError},
     prelude::*,
+    pybacked::PyBackedStr,
     types::PyDateTime,
 };
+use raphtory_api::core::entities::VID;
 use serde::Serialize;
 use std::{future::Future, thread};
 
 pub mod errors;
 pub(crate) mod export;
 mod module_helpers;
 
-/// Extract a `NodeRef` from a Python object.
-/// The object can be a `str`, `u64` or `PyNode`.
-/// If the object is a `PyNode`, the `NodeRef` is extracted from the `PyNode`.
-/// If the object is a `str`, the `NodeRef` is created from the `str`.
-/// If the object is a `int`, the `NodeRef` is created from the `int`.
-///
-/// Arguments
-///     vref: The Python object to extract the `NodeRef` from.
-///
-/// Returns
-///    A `NodeRef` extracted from the Python object.
-impl<'source> FromPyObject<'source> for NodeRef<'source> {
-    fn extract(vref: &'source PyAny) -> PyResult<Self> {
-        if let Ok(s) = vref.extract::<&'source str>() {
-            Ok(NodeRef::External(GidRef::Str(s)))
-        } else if let Ok(gid) = vref.extract::<u64>() {
-            Ok(NodeRef::External(GidRef::U64(gid)))
-        } else if let Ok(v) = vref.extract::<PyNode>() {
-            Ok(NodeRef::Internal(v.node.node))
+#[derive(Debug, Eq, PartialEq, Hash)]
+pub enum PyNodeRef {
+    ExternalStr(PyBackedStr),
+    ExternalInt(u64),
+    Internal(VID),
+}
+
+impl<'source> FromPyObject<'source> for PyNodeRef {
+    fn extract_bound(ob: &Bound<'source, PyAny>) -> PyResult<Self> {
+        if let Ok(s) = ob.extract::<PyBackedStr>() {
+            Ok(PyNodeRef::ExternalStr(s))
+        } else if let Ok(gid) = ob.extract::<u64>() {
+            Ok(PyNodeRef::ExternalInt(gid))
+        } else if let Ok(v) = ob.extract::<PyNode>() {
+            Ok(PyNodeRef::Internal(v.node.node))
         } else {
             Err(PyTypeError::new_err("Not a valid node"))
         }
     }
 }
 
+impl AsNodeRef for PyNodeRef {
+    fn as_node_ref(&self) -> NodeRef {
+        match self {
+            PyNodeRef::ExternalStr(str) => NodeRef::External(GidRef::Str(str)),
+            PyNodeRef::ExternalInt(gid) => NodeRef::External(GidRef::U64(*gid)),
+            PyNodeRef::Internal(vid) => NodeRef::Internal(*vid),
+        }
+    }
+}
+
+// TODO: Revisit once the two lifetime version of FromPyObject is available in pyo3 (see https://github.com/PyO3/pyo3/pull/4390)
```

Comment:
Wanna raise this as an issue in raphtory?

Reply:
Done: https://github.com/Pometry/Raphtory/issues/1850